morpheus/_lib/doca/src/doca_convert_stage.cpp

```diff
@@ -0,0 +1,296 @@
+{
+    auto packet_count      = raw_msg->count();
+    auto max_size          = raw_msg->get_max_size();
+    auto pkt_addr_list     = raw_msg->get_pkt_addr_list();
+    auto pkt_hdr_size_list = raw_msg->get_pkt_hdr_size_list();
+    auto pkt_pld_size_list = raw_msg->get_pkt_pld_size_list();
+    auto queue_idx         = raw_msg->get_queue_idx();
+
+    const auto payload_buff_size = doca::gather_sizes(packet_count, pkt_pld_size_list, m_stream_cpp);
+
+    const auto header_buff_size = packet_count * sizeof(uint32_t);
+    const auto sizes_buff_size  = packet_count * sizeof(uint32_t);
+
+    auto packet_buffer =
+        doca::PacketDataBuffer(packet_count, header_buff_size, payload_buff_size, sizes_buff_size, m_stream_cpp);
+
+    // gather payload data, intentionally calling this first as it needs to perform an early sync operation
+    doca::gather_payload(packet_count,
+                         pkt_addr_list,
+                         pkt_hdr_size_list,
+                         pkt_pld_size_list,
+                         static_cast<uint8_t*>(packet_buffer.m_payload_buffer->data()),
+                         m_stream_cpp);
+
+    // gather header data
+    doca::gather_src_ip(
+        packet_count, pkt_addr_list, static_cast<uint32_t*>(packet_buffer.m_header_buffer->data()), m_stream_cpp);
+
+    MRC_CHECK_CUDA(cudaMemcpyAsync(static_cast<uint8_t*>(packet_buffer.m_payload_sizes_buffer->data()),
+                                   pkt_pld_size_list,
+                                   sizes_buff_size,
+                                   cudaMemcpyDeviceToDevice,
+                                   m_stream_cpp));
+    cudaStreamSynchronize(m_stream_cpp);
+
+    m_buffer_channel->await_write(std::move(packet_buffer));
+}
+
+void DocaConvertStage::buffer_reader(rxcpp::subscriber<source_type_t>& output)
+{
+    std::vector<doca::PacketDataBuffer> packets;
+
+    while (!m_buffer_channel->is_channel_closed())
+    {
+        std::size_t ttl_packets             = 0;
+        std::size_t ttl_header_bytes        = 0;
+        std::size_t ttl_payload_bytes       = 0;
+        std::size_t ttl_payload_sizes_bytes = 0;
+        const auto poll_end                 = std::chrono::high_resolution_clock::now() + m_max_time_delta;
+        while (std::chrono::high_resolution_clock::now() < poll_end && !m_buffer_channel->is_channel_closed())
+        {
+            doca::PacketDataBuffer packet_buffer;
+            auto status = m_buffer_channel->await_read_until(packet_buffer, poll_end);
+
+            if (status == mrc::channel::Status::success)
+            {
+                ttl_packets += packet_buffer.m_num_packets;
+                ttl_header_bytes += packet_buffer.m_header_buffer->size();
+                ttl_payload_bytes += packet_buffer.m_payload_buffer->size();
+                ttl_payload_sizes_bytes += packet_buffer.m_payload_sizes_buffer->size();
+                packets.emplace_back(std::move(packet_buffer));
```

Comment:
Add a max batch size option which can break the loop when `ttl_packets` exceeds a certain size. The size should be configurable by the user.